datastore/storagebackend/postgresql/gettsattrcombos.go

```diff
@@ -0,0 +1,294 @@
+		if !found {
+			return nil, fmt.Errorf("col2attr: no value found for key >%s<", col)
+		}
+		attrs = append(attrs, attr)
+	}
+	return attrs, nil
+}
+
+// getTSAllCols returns names of all database columns.
+func getTSAllCols() ([]string, error) {
+	cols := []string{}
+	for col := range col2attr {
+		cols = append(cols, col)
+	}
+	return cols, nil
+}
+
+// getTSAttrCols returns names of database columns corresponding to attrs.
+func getTSAttrCols(attrs []string) ([]string, error) {
+	seen := map[string]struct{}{}
+	cols := []string{}
+
+	supAttrs := func() []string {
+		attrs := []string{}
+		for a := range attr2col {
+			attrs = append(attrs, a)
+		}
+		return attrs
+	}
+
+	for _, attr := range attrs {
+		col, found := attr2col[attr]
+		if !found { // not an attribute name
+			_, found0 := col2attr[attr]
+			if !found0 { // not a column name either
+				return nil, fmt.Errorf(
+					"attribute not found: %s; supported attributes: %s",
+					attr, strings.Join(supAttrs(), ", "))
+			}
+			col = attr // attr is already a column name
+		}
+
+		if _, found = seen[col]; found {
+			return nil, fmt.Errorf("attribute %s specified more than once", attr)
+		}
+
+		cols = append(cols, col)
+	}
+
+	return cols, nil
+}
+
+// getTSMdata returns a TSMetadata object initialized from colVals.
+func getTSMdata(colVals map[string]interface{}) (*datastore.TSMetadata, error) {
+
+	tsMData := datastore.TSMetadata{}
+	tp := reflect.ValueOf(&tsMData)
+
+	for col, val := range colVals {
+		attr, found := col2attr[col]
+		if !found {
+			return nil, fmt.Errorf(
+				"key not found in col2attr: %s (existing contents: %v)", col, col2attr)
+		}
+
+		field := tp.Elem().FieldByName(attr)
+		if !field.IsValid() {
+			return nil, fmt.Errorf("invalid field (attr: %s, col: %s)", attr, col)
+		}
+		if !field.CanSet() {
+			return nil, fmt.Errorf("unassignable field (attr: %s, col: %s)", attr, col)
+		}
+
+		switch field.Kind() {
+		case reflect.String:
+			val0, ok := val.(string)
+			if !ok {
+				return nil, fmt.Errorf(
+					"value not string: %v (type: %T; attr: %s, col: %s", val, val, attr, col)
+			}
+			field.SetString(val0)
+		default:
+			return nil, fmt.Errorf(
+				"unsupported type: %v (val: %v; type: %T; attr: %s; col: %s)",
+				field.Kind(), val, val, attr, col)
+		}
+	}
+
+	return &tsMData, nil
+}
+
+// scanTsMdata scans column cols from current result row in rows and converts to a
+// TSMetadata object.
+// Returns (TSMetadata object, nil) upon success, otherwise (..., error).
+func scanTsMdata(rows *sql.Rows, cols []string) (*datastore.TSMetadata, error) {
+	colVals0 := make([]interface{}, len(cols))   // column values
+	colValPtrs := make([]interface{}, len(cols)) // pointers to column values
+	for i := range colVals0 {
+		colValPtrs[i] = &colVals0[i]
+	}
+	// scan row into column value pointers
+	if err := rows.Scan(colValPtrs...); err != nil {
+		return nil, fmt.Errorf("rows.Scan() failed: %v", err)
+	}
+	// combine column names and -values into a map
```

Comment:
```suggestion
	// combine column names and values into a map
```

Reply:
Hm ... I always wondered how to do this in English. In Norwegian we would say: "kombinere kolonnenavn og -verdier" instead of "kombinere kolonnenavn og kolonneverdier", so the dash is used to avoid having to write the prefix ("kolonne") for each item after the first one. So I thought in English you could do the same thing by saying "combine column names and -values" instead of the more verbose "combine column names and column values". Strictly speaking, without the dash it wouldn't be clear whether "values" would be related to columns or not.

Reply:
Ah, I thought it was just a typo.
I don't know how to do that in English, but happy to keep the Norwegian convention in our code comments.

Reply:
Ok, let's keep it then. (Bit of a bike shed discussion, but it's good to have those once in a while :) ).